### PR TITLE
Factor SessionPool out of ConnectionImpl.

### PR DESCRIPTION
**Pull request by mr-salty**
Make `SessionPool` its own class to avoid clutter in `ConnectionImpl`
and to make it easier to test.

Since `SessionPool` is a member of `ConnectionImpl`, which guarantees
the former cannot outlive the latter; only `ConnectionImpl` is required
to be reference counted. This simplifies lifetime management, avoiding
some subtle issues and race conditions, at the cost of requiring some
ostensibly pool-related private methods in `ConnectionImpl` for the
benefit of `SessionHolder`

There are no functional changes compared to the current implementation.

Part of #307

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp-spanner/916)

### CODE REVIEW

**Comment by mr-salty** [comment]
___
*[google/cloud/spanner/internal/connection_impl.h, line 61 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp-spanner/916#-Lq9lsARBztlBV4T17k-:-Lq9lsARBztlBV4T17k0:bvei9lp) ([raw file](https://github.com/googleapis/google-cloud-cpp-spanner/blob/cd0c3bccb29e88b52b57656bd68851e22318260b/google/cloud/spanner/internal/connection_impl.h#L61)):*
> ```objc
>         DefaultConnectionBackoffPolicy());
> 
> // Interface used by the SessionPool to manage sessions
> ```

@devjgm @devbww I don't really like this especially given it leads to multiple inheritance, perhaps we can discuss alternatives tomorrow?

It will eventually get other methods like DeleteSession and GetSession/RefreshSession or similar

File: google/cloud/spanner/internal/connection_impl.h [hunk 1]
```diff
@@ -120,19 +128,35 @@ class ConnectionImpl : public Connection,
...
-  // The current session pool.
-  // TODO(#307) - improve session refresh and expiration.
-  std::mutex mu_;
-  std::vector<std::unique_ptr<Session>> sessions_;  // GUARDED_BY(mu_)
+  std::shared_ptr<SpannerStub> stub_;
 
   std::unique_ptr<RetryPolicy> retry_policy_;
   std::unique_ptr<BackoffPolicy> backoff_policy_;
+
+  friend class SessionPool;
```
Comment (devbww) [inline google/cloud/spanner/internal/connection_impl.h]: Why do we need this friendship?

File: google/cloud/spanner/internal/connection_impl.h [hunk 2]
```diff
@@ -56,12 +58,22 @@ std::shared_ptr<ConnectionImpl> MakeConnection(
     std::unique_ptr<BackoffPolicy> backoff_policy =
         DefaultConnectionBackoffPolicy());
 
+// Interface used by the SessionPool to manage sessions
+class SessionManager {
```
Comment (devbww) [inline google/cloud/spanner/internal/connection_impl.h]: Should we have a separate header for SessionManager?

File: google/cloud/spanner/internal/session_pool.cc [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+inline namespace SPANNER_CLIENT_NS {
+namespace internal {
+
+StatusOr<std::unique_ptr<Session>> SessionPool::Allocate() {
+  {
+    std::lock_guard<std::mutex> lk(mu_);
+    if (!sessions_.empty()) {
+      auto session = std::move(sessions_.back());
+      sessions_.pop_back();
+      return {std::move(session)};
```
Comment (devbww) [inline google/cloud/spanner/internal/session_pool.cc]: Are the `{}`s necessary?  Indeed, is the `move` necessary?

File: google/cloud/spanner/internal/session_pool.cc [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    }
+  }
+
+  auto sessions = manager_->CreateSessions(/*num_sessions=*/1);
+  if (!sessions.ok()) {
+    return std::move(sessions).status();
+  }
+  // TODO(#307) for now, assume CreateSessions() returns exactly one Session on
```
Comment (devbww) [inline google/cloud/spanner/internal/session_pool.cc]: Nit: Didn't we ask for exactly one?

File: google/cloud/spanner/internal/session_pool.h [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+  /**
+   * Allocate a session from the pool, creating a new `Session` if necessary.
+   * @returns an error if session creation fails; always returns a non-null
+   * pointer on success.
+   */
+  StatusOr<std::unique_ptr<Session>> Allocate();
+
+  /// Release `session` back to the pool.
+  void Release(Session* session);
```
Comment (devbww) [inline google/cloud/spanner/internal/session_pool.h]: Why the asymmetry on `unique_ptr` -v- raw pointer between `Allocate()` and `Release()`.

File: google/cloud/spanner/internal/session_pool.h [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+ private:
+  std::mutex mu_;
+  std::vector<std::unique_ptr<Session>> sessions_;  // GUARDED_BY(mu_)
```
Comment (devbww) [inline google/cloud/spanner/internal/session_pool.h]: Might this be a vector of dissociated `SessionHolder`s instead?  Having two different varieties of `unique_ptr<Session>` might be confusing.

File: google/cloud/spanner/internal/session_pool_test.cc [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+  EXPECT_EQ((*session)->session_name(), "session1");
+
+  auto session2 = pool.Allocate();
+  ASSERT_STATUS_OK(session2);
+  EXPECT_EQ((*session2)->session_name(), "session2");
+
+  pool.Release((*session).release());
+  pool.Release((*session2).release());
+
+  // We should get the sessions back in the reverse order they were released.
```
Comment (devbww) [inline google/cloud/spanner/internal/session_pool_test.cc]: Perhaps expound on why this is (currently?) important.

Comment (devbww) [reply google/cloud/spanner/internal/session_pool_test.cc]: > Done.

OK, although I'd still say it is unclear why this is a necessary/desirable property.

File: google/cloud/spanner/internal/session_pool.cc (hunk 3 above)
Comment (devbww) [reply google/cloud/spanner/internal/session_pool.cc]: It looks like Clang-3.8 and GCC-4.8 didn't like something there.  Sorry.

**Pull request merged**